activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -179,21 +179,30 @@ def clear(connection)
         end
       end
 
-      class LeaseRegistry # :nodoc:
-        def initialize
-          @mutex = Mutex.new
-          @map = WeakThreadKeyMap.new
+      if RUBY_ENGINE == "ruby"
+        # Thanks to the GVL, the LeaseRegistry doesn't need to be synchronized on MRI
+        class LeaseRegistry < WeakThreadKeyMap # :nodoc:
+          def [](context)
+            super || (self[context] = Lease.new)
```

Comment:
Is this safe just because we know that we _are_ `context`, and therefore even if we stall during `Lease.new`, we can't possibly be racing someone else?